From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wayne Chang <waynec@nvidia.com>
Date: Fri, 30 Aug 2024 07:04:39 +0000
Subject: [PATCH 150/241] NVIDIA: SAUCE: usb: gadget: tegra-xudc: Reduce
 ping.LFPS trepeat for U1 exit reliability

We observed that some hosts may fail to exit U1 due to a U1 ping timeout.
By adjusting the SSPX_CORE_CNT27_PING_LFPS_TRPT parameter to reduce
the ping.LFPS trepeat value, the time is reduced from 300ms to 200ms and
improve the reliability of U1 exit in these scenarios.

Bug 4758552

Change-Id: I0edd565c3666ededbd0c1c95f3a134ef52e817b8
Signed-off-by: Wayne Chang <waynec@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3250526
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: WK Tsai <wtsai@nvidia.com>
Reviewed-by: HaoTien Hsu <haotienh@nvidia.com>
Reviewed-by: Henry Lin <henryl@nvidia.com>
(cherry picked from commit 32bff2af996df989efdf3607c9e3576fc6346081)

diff --git a/drivers/usb/gadget/udc/tegra-xudc.c b/drivers/usb/gadget/udc/tegra-xudc.c
index 3ed44e8cc712..b5cf8a375d73 100644
--- a/drivers/usb/gadget/udc/tegra-xudc.c
+++ b/drivers/usb/gadget/udc/tegra-xudc.c
@@ -3548,6 +3548,15 @@ static void tegra264_xudc_device_params_init(struct tegra_xudc *xudc)
 	val = xudc_readl(xudc, BLCG) & ~BLCG_COREPLL_PWRDN;
 	xudc_writel(xudc, val, BLCG);
 
+#define SSPX_CORE_CNT27 0x407c
+#define  SSPX_CORE_CNT27_PING_LFPS_TRPT_MASK GENMASK(29, 0)
+#define  SSPX_CORE_CNT27_PING_LFPS_TRPT(x) ((x) & \
+					SSPX_CORE_CNT27_PING_LFPS_TRPT_MASK)
+	val = xudc_readl(xudc, SSPX_CORE_CNT27);
+	val &= ~(SSPX_CORE_CNT27_PING_LFPS_TRPT_MASK);
+	val |= SSPX_CORE_CNT27_PING_LFPS_TRPT(0x7a1200);
+	xudc_writel(xudc, val, SSPX_CORE_CNT27);
+
 #define SSPX_CORE_CNT32 0x4090
 	/*
 	 * Compliance suite appears to be violating polling
-- 
2.43.0

